### DISCUSSION TITLE
What's the best strategy for packaging pre-release artifacts?

### DISCUSSION THREAD

**Discussion by jrouly** [projen/projen#2759] [Q&A] [answered]
I've got a JSII project using Projen, and I'm curious about the correct workflow for packaging pre-release artifacts.

For example, if my project's current version is `0.0.1`, then on a given PR I'd like to publish a pre-release artifact (e.g. `0.0.1-dev.0`) in all of my packaging targets (e.g. Python). Ideally following the documented JSII [version constraints](https://aws.github.io/jsii/user-guides/lib-author/configuration/targets/python/#prerelease-versions).

If I simply run `npm run build ; npm run package` it will package the artifacts as `0.0.1` (current version). I've even tried setting the `PRERELEASE` environment variable with no effect.

And if I enable `release: true`  and include `npm run bump`, the version does get correctly set after bumping but its reset by the downstream tasks (`build` and `package`).

**Comment by mrgrain** [comment]
```
I've got a JSII project using Projen, and I'm curious about the correct workflow for packaging pre-release artifacts.
```

In theory it should be:
1. Version on the branch is `0.0.0`
2. `PRERELEASE=dev projen bump`
3. `CI=0 RELEASE=true projen build` -> this produces the artifacts in the `dist` directory
4. `projen umbump` -> only needed if you care that the repo is returned to a "clean" state

**Reply by jrouly** [reply]
So I ran into two issues with that.
1. I had to set `CI=''` because, on CI, `projen build` doesn't actually package anything.
2. While `projen bump` will properly bump the package version in `package.json`, `projen build` is unaware of this change and ends up building packages versioned at the unbumped version (`0.0.0`).

Side note. When I don't disable `CI` and instead explicitly call `projen package`, it also ignores the bumped package version.

**Reply by mrgrain** [reply]
Ah yes, you also need to set `RELEASE=true` environment variable, otherwise the version gets reset `0.0.0`

**Reply by jrouly** [reply]
```shell
export CI=
export PRERELEASE=dev
export RELEASE=true

npm ci
git fetch --tags
npx projen bump
npx projen build
npx projen unbump
```

even with that environment variable set, I'm still seeing jsii artifacts (e.g. python) built at `0.0.0`. doesn't seem to have had any effect.

**Reply by jrouly** [reply, marked as answer]
```
export CI=
export PRERELEASE=dev
export RELEASE=true

npm ci
git fetch --tags
npx projen release
```

Interestingly, using `projen release` does appear to work -- which just runs `projen bump ; projen build ; projen unbump` like you recommended.

Running the tasks independently doesn't work, but running them together in one projen task does work.

**Reply by mrgrain** [reply]
Thanks for investigating.

> Running the tasks independently doesn't work, but running them together in one projen task does work.

Interesting and a bit surprising. But shows how messy this is at the moment 🙈
